### PR TITLE
Ria 5599

### PR DESCRIPTION
**Pull request by JamieHMCTS**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RIA-5599

### Change description ###
- ChangeDirectionDueDatePreparer is the ABOUT_TO_SUBMIT handler that handles the very beginning of the event. Its role is to prepare a DynamicList to show in the dropdown list of the first screen, where the user selects the `Direction` they want to change. I didn't have the time to see if the DynamicList object could be simplified further, considering there's a CCD field called DynamicList, so it'll have to be congruent with it.
The DynamicList object's fields are:
    - Value `(the selected direction)`
    - List<Value> `(the list of directions)`

<< the Direction to be changed gets selected >>

- ChangeDirectionDueDateMidEvent is the MID_EVENT handler that sets up the following page where the direction is edited.
It extrapolates what direction was selected to be changed and finds it from the list of directions. It then takes the details from that direction and writes them in bailCase fields:
    - BAIL_DIRECTION_EDIT_EXPLANATION
    - BAIL_DIRECTION_EDIT_PARTIES
    - BAIL_DIRECTION_EDIT_DATE_DUE
    - BAIL_DIRECTION_EDIT_DATE_SENT

<< The edited details of a direction are sent to the next screen >>

- ChangeDirectionDueDateHandler is the ABOUT_TO_SUBMIT handler that finds the direction that is being edited and updates the DIRECTIONS collection with one where that direction has the updated details (+ the list of Previous Dates nested)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmation.java [hunk 1]
```diff
@@ -34,7 +34,7 @@ public PostSubmitCallbackResponse handle(
         postSubmitResponse.setConfirmationHeader("# You have ended the application");
         postSubmitResponse.setConfirmationBody(
             "#### What happens next\n\n"
-                + "A notification has been sent to all parties. "
+                + "A notification has been sent to all sendDirectionList. "
```
Comment (Marcellofabbri) [inline src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmation.java:37]: The word "parties" got accidentally swapped for "sendDirectionList"

File: src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/BailApplicationSubmittedConfirmation.java [hunk 2]
```diff
@@ -26,7 +26,7 @@ public PostSubmitCallbackResponse handle(Callback<BailCase> callback) {
 
         postSubmitResponse.setConfirmationBody(
             "### What happens next\n\n"
-            + "All parties will be notified that the application has been submitted."
+            + "All sendDirectionList will be notified that the application has been submitted."
```
Comment (Marcellofabbri) [inline src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/BailApplicationSubmittedConfirmation.java:29]: The word "parties" got accidentally swapped for "sendDirectionList".

File: src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmationTest.java [hunk 3]
```diff
@@ -62,7 +62,7 @@ void should_set_confirmation_header_body() {
         assertEquals("# You have ended the application", response.getConfirmationHeader().get());
         assertThat(response.getConfirmationBody().get()).contains("#### What happens next\n\n");
         assertThat(response.getConfirmationBody().get())
-            .contains("A notification has been sent to all parties. No further action is required.");
+            .contains("A notification has been sent to all sendDirectionList. No further action is required.");
```
Comment (Marcellofabbri) [inline src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmationTest.java:65]: The word "parties" got accidentally swapped for "sendDirectionList".

File: src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmation.java (hunk 1 above)
Comment (JamieHMCTS) [reply src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmation.java]: I have changed sendDirectionList back to parties

File: src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/BailApplicationSubmittedConfirmation.java (hunk 2 above)
Comment (JamieHMCTS) [reply src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/BailApplicationSubmittedConfirmation.java]: I have changed sendDirectionList back to parties

File: src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmationTest.java (hunk 3 above)
Comment (JamieHMCTS) [reply src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/postsubmit/EndApplicationConfirmationTest.java]: I have changed sendDirectionList back to parties

File: src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueMidEvent.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+import uk.gov.hmcts.reform.bailcaseapi.domain.entities.ccd.field.IdValue;
+import uk.gov.hmcts.reform.bailcaseapi.domain.handlers.PreSubmitCallbackHandler;
+
+@Component
+public class ChangeDirectionDueMidEvent implements PreSubmitCallbackHandler<BailCase> {
+
+    private static final String direction = "Direction ";
```
Comment (neha-aggarw) [inline src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueMidEvent.java:28]: Please make the final variables all uppercase.

File: src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueMidEvent.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            .orElseThrow(() -> new IllegalStateException("bailDirectionList is missing"));
+
+        maybeDirections
+            .orElse(emptyList())
+            .stream()
+            .filter(idValue -> bailDirectionList.getValue().getCode().contains(
+                direction + (maybeDirections.orElse(emptyList()).size() - (Integer.parseInt(idValue.getId())) + 1)))
+            .forEach(idValue -> {
+
+                bailCase.write(BailCaseFieldDefinition.BAIL_DIRECTION_EDIT_EXPLANATION,
```
Comment (neha-aggarw) [inline src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueMidEvent.java:66]: The ccd_definitions here will hold the values related to last direction only. Can you please check the logic here?

File: src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        when(bailCase.read(BAIL_DIRECTION_LIST,DynamicList.class)).thenReturn(Optional.of(new DynamicList(direction1)));
+        when(bailCase.read(BAIL_DIRECTION_EDIT_DATE_DUE, String.class)).thenReturn(Optional.of("2222-12-01"));
+
+        PreSubmitCallbackResponse<BailCase> callbackResponse =
+            changeDirectionDueDateHandler.handle(PreSubmitCallbackStage.ABOUT_TO_SUBMIT, callback);
+
+        assertNotNull(callbackResponse);
+        assertEquals(bailCase, callbackResponse.getData());
+
+        //problem is here
```
Comment (neha-aggarw) [inline src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java:130]: Do we need this comment?

File: src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        assertThatThrownBy(() -> changeDirectionDueDateHandler.handle(null, callback))
+            .hasMessage("callbackStage must not be null")
+            .isExactlyInstanceOf(NullPointerException.class);
+
+        assertThatThrownBy(() -> changeDirectionDueDateHandler.handle(PreSubmitCallbackStage.ABOUT_TO_SUBMIT, null))
+            .hasMessage("callback must not be null")
+            .isExactlyInstanceOf(NullPointerException.class);
+    }
+
+    // remove when new CCD definitions are in Prod
```
Comment (neha-aggarw) [inline src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java:219]: Please check the classes and tests. We don't have any outdated ccd definitions and we should be working on what works with our set of definitions.

**Review by neha-aggarw — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        when(bailCase.read(DIRECTIONS)).thenReturn(Optional.of(existingDirections));
+
+        // Direction 2 is selected to be changed (first field of DynamicList)
+        DynamicList dynamicList = new DynamicList(new Value("Direction 2", "Direction 2"),
+                                                  List.of(
+                                                      new Value("Direction 2", "Direction 2"),
+                                                      new Value("Direction 1", "Direction 1")));
+
+
+        // "Direction 1" in UI is equivalent of Direction with IdValue "2" in backend
```
Comment (neha-aggarw) [inline src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java:113]: Is this comment right? based on the message Marcello put in chat?
Direction 1 -> IdValue 4 -> index 0
Direction 2 -> IdValue 3 -> index 1

**Review by neha-aggarw — APPROVED** [review]

File: src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java (hunk 8 above)
Comment (Marcellofabbri) [reply src/test/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/ChangeDirectionDueDateHandlerTest.java:113]: If you have 4 directions yes:
```
Direction 1 -> IdValue 4 -> index 0
Direction 2 -> IdValue 3 -> index 1
Direction 3 -> IdValue 2 -> index 2
Direction 4 -> IdValue 1 -> index 3
```

In the test there are two only:
```
Direction 1 -> IdValue 2 -> index 0
Direction 2 -> IdValue 1 -> index 1
```

Basically the Directions are ordered in the direction from most recent to oldest. 
The direction number is from 1 onwards.
The index in that collection is from 0 onwards.
The Idvalue is incremental but from oldest to newest, so if we have 4 directions it'll be 4, if we have 6 directions it'll be 6.

If you have 6 directions:
```
Direction 1 -> IdValue 6 -> index 0
Direction 2 -> IdValue 5 -> index 1
Direction 3 -> IdValue 4 -> index 2
Direction 4 -> IdValue 3 -> index 3
Direction 5 -> IdValue 2 -> index 4
Direction 6 -> IdValue 1 -> index 5
```

**Review by Dbarnes-v1 — APPROVED** [review]

**Pull request merged**